aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1561,7 +1561,9 @@ std::tuple<Tensor, Tensor, Tensor> _lu_with_info(const Tensor& self, bool comput
   auto req_size = self.sizes().vec();
   req_size.pop_back();
   req_size.back() = std::min(m, n);
-  auto pivots_tensor = at::empty(req_size, self.options().dtype(kInt));
+  auto k = std::min(m, n);
+  // use arange since cuSOLVER does not return the correct values for ill-conditioned matrices.
+  auto pivots_tensor = at::arange(1, k+1, self.options().dtype(kInt)).expand(req_size).contiguous();
```

Comment:
I wonder how much of the total execution time this arange takes. Before this PR arange was called only for non-pivoted MAGMA functions, now it's called for every invocation of the function.
If we have the problem only with cuSOLVER, then maybe we should do it only in the `lu_cusolver_looped` function.

Reply:
Perf results:
```
[-------------------------- LU factorization (ATen) torch.float64 ---------------------------]
                                     |  lu_factorize ARANGE PIVOT  |  lu_factorize EMPTY PIVOT
1 threads: -----------------------------------------------------------------------------------
      torch.Size([1, 1, 1])          |              265.2          |             227.4        
      torch.Size([2, 1, 1])          |              311.6          |             222.2        
      torch.Size([4, 1, 1])          |              337.1          |             246.8        
      torch.Size([8, 1, 1])          |              389.1          |             295.9        
      torch.Size([16, 1, 1])         |              549.6          |             462.7        
      torch.Size([32, 1, 1])         |              532.0          |             444.3        
      torch.Size([64, 1, 1])         |              534.0          |             440.2        
      torch.Size([128, 1, 1])        |              532.5          |             444.7        
      torch.Size([1, 2, 2])          |              257.8          |             217.2        
      torch.Size([2, 2, 2])          |              319.8          |             230.0        
      torch.Size([4, 2, 2])          |              346.2          |             256.0        
      torch.Size([8, 2, 2])          |              396.7          |             305.5        
      torch.Size([16, 2, 2])         |              538.2          |             448.3        
      torch.Size([32, 2, 2])         |              535.8          |             447.4        
      torch.Size([64, 2, 2])         |              535.4          |             447.1        
      torch.Size([128, 2, 2])        |              535.1          |             448.3        
      torch.Size([1, 8, 8])          |              253.5          |             216.9        
      torch.Size([2, 8, 8])          |              313.5          |             231.2        
      torch.Size([4, 8, 8])          |              377.3          |             286.4        
      torch.Size([8, 8, 8])          |              425.9          |             340.1        
      torch.Size([16, 8, 8])         |              536.3          |             449.2        
      torch.Size([32, 8, 8])         |              536.0          |             453.1        
      torch.Size([64, 8, 8])         |              538.2          |             453.1        
      torch.Size([128, 8, 8])        |              540.2          |             454.2        
      torch.Size([1, 16, 16])        |              262.5          |             229.8        
      torch.Size([2, 16, 16])        |              366.9          |             280.9        
      torch.Size([4, 16, 16])        |              427.4          |             329.9        
      torch.Size([8, 16, 16])        |              567.1          |             478.6        
      torch.Size([16, 16, 16])       |              552.1          |             456.4        
      torch.Size([32, 16, 16])       |              547.9          |             465.4        
      torch.Size([64, 16, 16])       |              550.7          |             466.9        
      torch.Size([128, 16, 16])      |              555.0          |             470.9        
      torch.Size([1, 32, 32])        |              338.8          |             288.6        
      torch.Size([2, 32, 32])        |              587.0          |             493.8        
      torch.Size([4, 32, 32])        |              587.0          |             501.3        
      torch.Size([8, 32, 32])        |              586.2          |             501.9        
      torch.Size([16, 32, 32])       |              587.7          |             502.5        
      torch.Size([32, 32, 32])       |              595.2          |             512.2        
      torch.Size([64, 32, 32])       |              605.4          |             524.7        
      torch.Size([128, 32, 32])      |              660.6          |             571.8        
      torch.Size([1, 64, 64])        |              467.6          |             432.2        
      torch.Size([2, 64, 64])        |              772.5          |             688.5        
      torch.Size([4, 64, 64])        |              773.0          |             690.1        
      torch.Size([8, 64, 64])        |              775.1          |             692.8        
      torch.Size([16, 64, 64])       |              810.3          |             713.3        
      torch.Size([32, 64, 64])       |              812.7          |             728.4        
      torch.Size([64, 64, 64])       |              875.4          |             793.2        
      torch.Size([128, 64, 64])      |             1007.9          |             926.1        
      torch.Size([1, 128, 128])      |              808.7          |             776.4        
      torch.Size([2, 128, 128])      |             1106.3          |            1024.5        
      torch.Size([4, 128, 128])      |             1120.7          |            1037.0        
      torch.Size([8, 128, 128])      |             1149.2          |            1055.6        
      torch.Size([16, 128, 128])     |             1223.6          |            1129.3        
      torch.Size([32, 128, 128])     |             1400.6          |            1309.6        
      torch.Size([64, 128, 128])     |             1733.2          |            1644.2        
      torch.Size([128, 128, 128])    |             2491.8          |            2392.1        
      torch.Size([1, 512, 512])      |             4470.5          |            4434.9        
      torch.Size([2, 512, 512])      |             4749.8          |            4646.0        
      torch.Size([4, 512, 512])      |             5637.3          |            5525.7        
      torch.Size([8, 512, 512])      |             7503.4          |            7373.8        
      torch.Size([16, 512, 512])     |            11333.7          |           11219.3        
      torch.Size([32, 512, 512])     |            20525.1          |           20884.9        
      torch.Size([64, 512, 512])     |            37731.9          |           37597.3        
      torch.Size([128, 512, 512])    |            71898.3          |           71811.9        
      torch.Size([1, 1024, 1024])    |            13294.3          |           13261.0        
      torch.Size([2, 1024, 1024])    |            19890.1          |           19716.2        
      torch.Size([4, 1024, 1024])    |            27530.9          |           27296.2        
      torch.Size([8, 1024, 1024])    |            42812.2          |           42476.9        
      torch.Size([16, 1024, 1024])   |            72920.9          |           72787.4        
      torch.Size([32, 1024, 1024])   |           138562.5          |          138553.8        
      torch.Size([64, 1024, 1024])   |           267677.2          |          265333.8        
      torch.Size([128, 1024, 1024])  |           524534.1          |          520628.5        

Times are in microseconds (us).
```

The arange makes it a few 10s of ms slower for the smaller use cases. I think it would be best to change it only for cusolver.  Let me check if I can directly init the ranged tensor within cuSOLVER without needing to change the API for all the LU calls.

Reply:
Reverted to empty and then init arange in the cusolver function.